[JBTM-3524] LRA records to be listed as part of ObjStoreBrowser

ochaloup opened the pull request:
https://issues.redhat.com/browse/JBTM-3524

Implementation of the LRA records to be listed within JMX MBean records.

LRA MAIN JDK11
!QA_JTA !QA_JTS_JDKORB !QA_JTS_OPENJDKORB !QA_JTS_JACORB !BLACKTIE !XTS !PERF NO_WIN !RTS !AS_TESTS !TOMCAT !JACOCO

ochaloup commented:
@mmusgrov I'm labeling this as `Hold` as I would like ask you for your feedback first. But I think the listing works fine, in general.
`
One generic doubt which I have is if the LRA listing should be part of the `jboss.jta:type=ObjectStore` bean or if there should be registered in another bean like `jboss.lra:type=ObjectStore`?
(https://github.com/jbosstm/narayana/blob/master/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/tools/osb/mbean/ObjStoreBrowser.java#L149)

mmusgrov commented on rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LongRunningAction.java:
```diff
@@ -106,6 +106,11 @@ public LongRunningAction(LRAService lraService, Uid rcvUid) {
 
+    // used for MBean LRA listing, see com.arjuna.ats.arjuna.tools.osb.mbean.ObjStoreBrowser
+    public LongRunningAction(Uid rcvUid) {
+        this(null, rcvUid);
```
The variable `lraService` is expected to be non null so better is to just construct one:

``` java
this(new LRAService(), rcvUid);
```

ochaloup replied:
+1, thank you

mmusgrov commented on rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/tools/osb/mbean/LRAParticipantRecordWrapper.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    @Override
+    public String getParticipantPath() {
```
+1 (this path includes all of the other paths: status, complete, forget, etc)

mmusgrov commented on rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/tools/osb/mbean/ObjStoreBrowserLRATest.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    @Test
+    public void lraMBean() throws Exception {
```
Remark: it is a good test. In the future we should add testing for the `StateManager/BasicAction/LongRunningAction/Failed` record type.

mmusgrov commented on rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/tools/osb/mbean/ObjStoreBrowserLRATest.java:
```diff
@@ -0,0 +1,93 @@
+                    participantUrl + "/compensate", lraRecord.getCompensator());
+        } finally {
+            if (jmxMBean != null) {
+                jmxMBean.remove(true);
```
This works but why didn't you finish the LRA, and then re-probe, and then check that the beans for the LRA and participant are no longer present.

mmusgrov reviewed (changes requested):
Sorry I missed this (the Hold label threw me).

I just had a single change request about `LongRunningAction.java` where I have requested that you initialise the `lraService` variable.

mmusgrov commented on ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/tools/osb/mbean/ObjStoreBrowser.java:
```diff
@@ -140,13 +140,29 @@
     // The format is OSType1=BeanType1,OSType2=BeanType2,etc
     public static final String OBJ_STORE_BROWSER_HANDLERS = "com.arjuna.ats.arjuna.tools.osb.mbean.ObjStoreBrowserHandlers";
-    private static final String STORE_MBEAN_NAME = "jboss.jta:type=ObjectStore";
+    public static final String STORE_MBEAN_NAME = "narayana.logStore:type=ObjectStore";
```
Ah wait, I see that you override the name in ObjStoreJtaBrowser. Let me continue looking through the changes ...

mmusgrov commented:
> I did changes based on discussion at the Google Chat that I changed the bean name that will be exposing the object store to be `narayana.logStore:type=ObjectStore`. This needs to be changed within WildFly integration.

@ochaloup to clarify my understanding, is the following statement correct: the JMX MBeans for JTA will still use the domain name `jboss.jta` but the new JMX mbeans relating to LRA will use the domain name `narayana.logStore`?

mmusgrov reviewed (changes requested):
If I understand correctly the purpose of  the class`ObjStoreJtaBrowser` is to instrument JMX objects for backwards compatibility and it does this by keeping the old JMX domain name. If this is correct then it would be better to just add a method to the browser to configure the domain name, hence I am requesting either a change or further discussion/clarification.

ochaloup commented:
@mmusgrov I updated the PR with having the configuration being done in config bean. On top I think the LRA types should not be loaded as default types as LRA is only optional functionality to default JTA handling. I consider to add them separately in app runtime.

ochaloup commented:
A mistake by working late. I forgot to update the changed APIs in the tests. Should be fixed now.

mmusgrov commented on ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/tools/osb/mbean/ObjStoreBrowser.java:
```diff
@@ -198,44 +204,88 @@ private void registerMBeans() {
+     * @param typeName the type name handler that should be de-registered from handling by browser
+     * @return true if the type name was installed and now removed, false otherwise
+     */
+    public boolean removeType(String typeName) {
```
I'm not sure we want to allow the instrumentation types to be dynamic unless there is a requirement for it. Typically users create/develop new instrumentation types and then set them on the browser, ie the types are defined up front and not dynamically.

How would you imagine customers might use this feature?

ochaloup replied:
The reason why added it here is the WildFly integration. When the LRA service is stopped or redefined there was lack of the "remove" functionality. So from the "customer" perspective, this is used in WildFly integration part on service start and stop.

I do understand the perspective about the types are not considered to be dynamic and WildFly integration should be fine to reinstall only on JVM restart.

Would you think to not adding `removeType` functionality at all? I can remove it from the PR.

mmusgrov replied:
It is the JMX MBeans that need removing (see method `private void unregisterMBeans()`) .

If the LRA subsystem is removed then:
- would that not involve a reboot (in which case the WFLY subsystem would not register those LRA type handlers); 
- there should be no more new instances to instrument;
- what does the user do with the existing records left in the object store?

ochaloup replied:
Sure, no problem, I removed the method.

mmusgrov commented on ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/tools/osb/mbean/ObjStoreBrowser.java:
```diff
@@ -315,8 +369,8 @@ public UidWrapper findUid(Uid uid) {
     /**
-     * Find the registered beand corresponding to a uid.
-     * @deprecated use {@link #findUid(com.arjuna.ats.arjuna.common.Uid)} ()} instead.
+     * Find the registered beand corresponding to an uid.
```
`a uid` is the correct usage of the indefinite article since it precedes a consonant.

ochaloup replied:
I see, the spellchecker told me opposite. Adjusted it back, thanks!

mmusgrov replied:
The idea spell checker says the same but it it is incorrect, `a uid` is the correct indefinite article to use.

mmusgrov commented on ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/tools/osb/mbean/ObjStoreBrowser.java:
```diff
@@ -526,7 +581,7 @@ public static String canonicalType(String type) {
                         String typeName = canonicalType(types.unpackString());
 
-                        if (typeName.length() == 0)
+                        if (typeName == null || typeName.length() == 0)
```
`types.unpackString()` can never be null unless someone packed a zero length string in the object store, why would someone do that? If you think we should protect against this eventuality then something is corrupt or boken so you may as well throw an ObjectStoreException and handle it in the catch block (and change the log message from trace to warn).

ochaloup replied:
I see. This change came was rather from my lack of understanding. I took it from the fact that method `canonicalType` may return the `null` and thus it seemed to me that the method considers the null type string as a possible option.
I will remove my adjustment as it's not necessary.

mmusgrov commented on ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/tools/osb/mbean/UidWrapper.java:
```diff
@@ -53,11 +54,13 @@ public static String getRecordWrapperTypeName() {
+	private final ObjStoreBrowser browser;
+	private final String browserMBeanName;
+	private final String beantype;
+    private final String className;
```
misaligned indent

mmusgrov replied:
You have replaced spaces with tabs in this file (I thought our style checker detected these issues?).

ochaloup replied:
I did nothing with indentation.  My point was to add the `final` identifier.
The trouble is that the file itself was wrongly indented before I did the changes here. Some lines were with space (the one with `className`) the others were with tabs.

Our style checker was permitted to be run only on `rts` and `cdi` modules. I got no permission to run it at other places.

What should I do here? Should I change all changed lines to have tab indentation? Or spaces?

mmusgrov replied:
If the tab was already present then no problem.

mmusgrov commented on ArjunaCore/arjuna/tests/classes/com/hp/mwtests/ts/arjuna/tools/ExposeAllLogsTest.java:
```diff
@@ -47,19 +47,18 @@
 import com.arjuna.ats.internal.arjuna.common.UidHelper;
 import com.arjuna.common.internal.util.propertyservice.BeanPopulator;
 
+
```
Are two blank lines after the import section deliberate?

ochaloup replied:
Not sure, will change it to one line.

mmusgrov commented on ArjunaCore/arjuna/tests/classes/com/hp/mwtests/ts/arjuna/tools/ExposeAllLogsTest.java:
```diff
@@ -126,17 +125,13 @@ private void test(boolean exposeAllLogsViaJMX) throws Exception
 
     private void probeObjectStore(boolean exposeAllLogs, boolean useJMX) throws MBeanException {
         ObjectStoreEnvironmentBean osEnvBean = BeanPopulator.getDefaultInstance(ObjectStoreEnvironmentBean.class);
-
```
You reformatted the code but now there is no separation between declarations and code and between control statements.

ochaloup replied:
Ok, I will return the lime back to as they were defined.

mmusgrov commented on ArjunaJTS/jtax/tests/classes/com/hp/mwtests/ts/jta/jts/tools/JTSObjStoreBrowserTest.java:
```diff
@@ -230,11 +230,11 @@ public void jtsMBeanTest() throws Exception {
         }
 
-        createObjStoreBrowser(true); // start an mbean server and object store browser
+		ObjStoreBrowser osb = createObjStoreBrowser(true); // start an mbean server and object store browser
```
You replaces spaces with tabs resulting in the code no longer lining uo

ochaloup replied:
+1, strange IDE behaviour, probably because part of the file is tab indented. I will fix it.

mmusgrov reviewed (changes requested):
The PR is mostly okay:
- I had some quibbles about the changes to the existing styling/grammar;
- and that you allow removal of instrumentation types at runtime

ochaloup commented:
@mmusgrov thanks for you review. I adjusted the code appropriately. I have two things for confirmation. I explained why I added the `removeType` method. I can remove it from this PR I would like to know what you think about my perspective on it.

The second minor thing is about indentation of one place in code. I would like know how to fix it for it's fine from your perspective.

ochaloup commented:
@mmusgrov I removed the method we were discussing about. 

The tests passed. The intermittent failure was caused by a trouble to write to object store which seems to be env issue not related to this PR.

```
Sep 10, 2021 1:02:55 PM com.arjuna.ats.arjuna.coordinator.TransactionReaper check
WARN: ARJUNA012378: ReaperElement appears to be wedged: java.io.FileDescriptor.sync(Native Method)
  com.arjuna.ats.internal.arjuna.objectstore.ShadowingStore.write_state(ShadowingStore.java:602)
  com.arjuna.ats.internal.arjuna.objectstore.FileSystemStore.write_state_internal(FileSystemStore.java:365)
  com.arjuna.ats.internal.arjuna.objectstore.FileSystemStore.write_committed(FileSystemStore.java:139)
  com.arjuna.ats.arjuna.coordinator.BasicAction.deactivate(BasicAction.java:550)
  io.narayana.lra.coordinator.domain.model.LongRunningAction.updateState(LongRunningAction.java:558)
  io.narayana.lra.coordinator.domain.model.LongRunningAction.abortLRA(LongRunningAction.java:931)
  io.narayana.lra.coordinator.domain.model.LongRunningAction.cancel(LongRunningAction.java:1021)
  com.arjuna.ats.arjuna.coordinator.TransactionReaper.doWork(TransactionReaper.java:431)
  com.arjuna.ats.internal.arjuna.coordinator.ReaperWorkerThread.run(ReaperWorkerThread.java:78)
```

mmusgrov commented:
> The tests passed. The intermittent failure was caused by a trouble to write to object store which seems to be env issue not related to this PR.
> 
> ```
> Sep 10, 2021 1:02:55 PM com.arjuna.ats.arjuna.coordinator.TransactionReaper check
> WARN: ARJUNA012378: ReaperElement appears to be wedged: java.io.FileDescriptor.sync(Native Method)
>   com.arjuna.ats.internal.arjuna.objectstore.ShadowingStore.write_state(ShadowingStore.java:602)
>   com.arjuna.ats.internal.arjuna.objectstore.FileSystemStore.write_state_internal(FileSystemStore.java:365)
>   com.arjuna.ats.internal.arjuna.objectstore.FileSystemStore.write_committed(FileSystemStore.java:139)
>   com.arjuna.ats.arjuna.coordinator.BasicAction.deactivate(BasicAction.java:550)
>   io.narayana.lra.coordinator.domain.model.LongRunningAction.updateState(LongRunningAction.java:558)
>   io.narayana.lra.coordinator.domain.model.LongRunningAction.abortLRA(LongRunningAction.java:931)
>   io.narayana.lra.coordinator.domain.model.LongRunningAction.cancel(LongRunningAction.java:1021)
>   com.arjuna.ats.arjuna.coordinator.TransactionReaper.doWork(TransactionReaper.java:431)
>   com.arjuna.ats.internal.arjuna.coordinator.ReaperWorkerThread.run(ReaperWorkerThread.java:78)
> ```

I created issue https://issues.redhat.com/browse/JBTM-3536 to track the test fix.
